hy/core/macros.hy

```diff
@@ -20,6 +20,20 @@
          ~@(interleave (repeat name) rest))
      ~name))
 
+
+(defmacro assoc [coll k1 v1 &rest other-kvs]
+  (if (odd? (len other-kvs))
+    (macro-error (last other-kvs) "`assoc` takes an odd number of arguments"))
+  (setv c (if other-kvs (gensym "c") coll))
+  `(setv
+    ~@(if other-kvs [c coll] [])
+    ~@(sum
+      (list-comp [`(get ~c ~k) v] [[k v] (zip
+        (+ (, k1) (cut other-kvs 0 None 2))
+        (+ (, v1) (cut other-kvs 1 None 2)))])
+      [])))
```

Comment:
Hmm, nope. A cons onto a generic iterable doesn't work like that for some reason. It would have to be `(chain [[k1 v1]] (partition other-kvs))`, or spliced like
```Hy
~@(sum (list-comp [`(get ~c ~k) v] [[k v] `(~[k1 v1] ~@(partition other-kvs))])
```